### PR TITLE
Sort pipelines

### PR DESCRIPTION
**Pull request by camilabarca**

### CODE REVIEW

File: ui/projects/etl-server/src/lib/pipelines/pipeline-list-item/pipeline-list-item.component.ts [hunk 1]
```diff
@@ -25,6 +25,7 @@ export class PipelineListItemComponent implements OnInit {
   }
 
   ngOnInit() {
+    // console.log(this.item);
```
Comment (akariv) [inline ui/projects/etl-server/src/lib/pipelines/pipeline-list-item/pipeline-list-item.component.ts:28]: Remove commented out code

File: ui/projects/etl-server/src/lib/pipelines/pipeline-list/pipeline-list.component.html [hunk 2]
```diff
@@ -1,6 +1,15 @@
 <app-container>
   <h1 class='theme-primary-fg' i18n>Processing Tasks</h1>
   <div class='controls'>
+    <div *ngIf='roles._.pipelinesNew'>
```
Comment (akariv) [inline ui/projects/etl-server/src/lib/pipelines/pipeline-list/pipeline-list.component.html:4]: The roles object details what the user is allowed to do.
We can allow all users to sort the pipelines, so there's no need in the `*ngIf` here

File: ui/projects/etl-server/src/lib/pipelines/pipeline-list/pipeline-list.component.html [hunk 3]
```diff
@@ -1,6 +1,15 @@
...
+      <mat-form-field appearance="fill">
+        <mat-label>Sort</mat-label>
+        <select (change) = 'selectChangeHandler($event)'>
```
Comment (akariv) [inline ui/projects/etl-server/src/lib/pipelines/pipeline-list/pipeline-list.component.html:7]: Use more descriptive name for function -
instead of `selectChangeHandler` you could use for example `sortByChanged`

File: ui/projects/etl-server/src/lib/pipelines/pipeline-list/pipeline-list.component.ts [hunk 4]
```diff
@@ -11,6 +11,8 @@ import { RolesService } from '../../roles.service';
 })
 export class PipelineListComponent implements OnInit {
 
+  selectedOption: string = '';
```
Comment (akariv) [inline ui/projects/etl-server/src/lib/pipelines/pipeline-list/pipeline-list.component.ts:14]: use better name for variable - `sortByField`, for example

File: ui/projects/etl-server/src/lib/pipelines/edit-pipeline/edit-pipeline.component.ts [hunk 5]
```diff
@@ -46,6 +46,8 @@ export class EditPipelineComponent implements OnInit {
       this.item.params = this.item.params || {};
       this.item.schedule = this.item.schedule || '';
       this.item.kind = this.item.kind || '';
+      const d = new Date();
```
Comment (akariv) [inline ui/projects/etl-server/src/lib/pipelines/edit-pipeline/edit-pipeline.component.ts:49]: Why is this here?

File: ui/projects/etl-server/src/lib/pipelines/pipeline-list/pipeline-list.component.ts [hunk 6]
```diff
@@ -62,5 +65,16 @@ export class PipelineListComponent implements OnInit {
     });
   }
 
+  selectChangeHandler (event: any) {
+    //update the ui
+    this.selectedOption = event.target.value;
+    if (this.selectedOption === 'created-date-up'){
```
Comment (akariv) [inline ui/projects/etl-server/src/lib/pipelines/pipeline-list/pipeline-list.component.ts:71]: you should sort the original pipelines and rebuild the sections instead of sorting the sections themselves.

**Review by akariv — CHANGES_REQUESTED** [review]
@camilabarca - left some review comments, thanks!

File: etl_server/pipelines/controllers.py [hunk 7]
```diff
@@ -139,8 +140,8 @@ def query_pipelines(self, user=None, public=None):
             status = dict(status='didnt-run')
             status = statuses.get(res['key'], status)
             res['value']['status'] = status
-        query_results['result'] = list(map(lambda x: dict(x.get('value'), owner=x.get('owner'), private=x.get('private')), results))
-
+        query_results['result'] = list(map(lambda x: dict(x.get('value'), owner=x.get('owner'), private=x.get('private'), 
+                                    created_at=(x['created_at']).isoformat(), updated_at=(x.get('updated_at')).isoformat()), results))
```
Comment (akariv) [inline etl_server/pipelines/controllers.py:144]: Why the difference between the two properties? For one you're using square brackets [] and for the other you're using get().
Also it would probably be better to check that the value is not None before calling isoformat() on it.

File: etl_server/pipelines/controllers.py [hunk 8]
```diff
@@ -127,6 +127,7 @@ def __get_logs(self, id):
             return '', ''
 
     def query_pipelines(self, user=None, public=None):
+        # assert(False)
```
Comment (akariv) [inline etl_server/pipelines/controllers.py:130]: Remove commented out code

File: ui/projects/etl-server/src/lib/pipelines/pipeline-list/pipeline-list.component.html [hunk 9]
```diff
@@ -1,6 +1,17 @@
 <app-container>
   <h1 class='theme-primary-fg' i18n>Processing Tasks</h1>
   <div class='controls'>
+    <div>
+      <mat-form-field appearance="fill">
+        <mat-label>Sort</mat-label>
+        <select [(ngModel)]='selected'>
+          <option value = ''></option>
```
Comment (akariv) [inline ui/projects/etl-server/src/lib/pipelines/pipeline-list/pipeline-list.component.html:8]: Let's have a default value here and not an empty selection

File: ui/projects/etl-server/src/lib/pipelines/pipeline-list/pipeline-list.component.ts [hunk 10]
```diff
@@ -11,6 +11,8 @@ import { RolesService } from '../../roles.service';
 })
 export class PipelineListComponent implements OnInit {
 
+  _selected = '';
```
Comment (akariv) [inline ui/projects/etl-server/src/lib/pipelines/pipeline-list/pipeline-list.component.ts:14]: Pleas use meaningful variable names (e.g. sortByField)

File: ui/projects/etl-server/src/lib/pipelines/pipeline-list/pipeline-list.component.ts [hunk 11]
```diff
@@ -47,11 +49,46 @@ export class PipelineListComponent implements OnInit {
         for (const pipeline of pipelines) {
           pipeline.display = pipeline.name;
         }
+    
         this.pipelineSections = this.processSections(pipelines, 0);
       }
     });
     this.api.queryPipelines();
   }
+  get selected() {
```
Comment (akariv) [inline ui/projects/etl-server/src/lib/pipelines/pipeline-list/pipeline-list.component.ts:58]: Meaningful names...

File: ui/projects/etl-server/src/lib/pipelines/pipeline-list/pipeline-list.component.ts [hunk 12]
```diff
@@ -47,11 +49,46 @@ export class PipelineListComponent implements OnInit {
...
+    return this._selected;
+  }
+
+  set selected(value) {
```
Comment (akariv) [inline ui/projects/etl-server/src/lib/pipelines/pipeline-list/pipeline-list.component.ts:62]: 1. This is a lot of code duplication
2. You're not using the sort by variable
3. There's no need to call the API again, as we already have the list of pipelines here...

**Review by akariv — CHANGES_REQUESTED** [review]

File: ui/projects/etl-server/src/lib/pipelines/pipeline-list/pipeline-list.component.ts (hunk 12 above)
Comment (camilabarca) [reply ui/projects/etl-server/src/lib/pipelines/pipeline-list/pipeline-list.component.ts]: what sort by variable do you mean?

File: etl_server/pipelines/controllers.py [hunk 13]
```diff
@@ -141,7 +141,8 @@ def query_pipelines(self, user=None, public=None):
             status = statuses.get(res['key'], status)
             res['value']['status'] = status
         query_results['result'] = list(map(lambda x: dict(x.get('value'), owner=x.get('owner'), private=x.get('private'), 
-                                    created_at=(x['created_at']).isoformat(), updated_at=(x.get('updated_at')).isoformat()), results))
+                                    created_at=((x.get('created_at')).isoformat() if x.get('created_at') != None else x.get('created_at')),
```
Comment (akariv) [inline etl_server/pipelines/controllers.py:143]: 1. In Python it's more correct to use `is not None` instead of `!= None`
2. If it is `None` you can use `None` in the else clause instead of `x.get(...)`

File: ui/projects/etl-server/src/lib/pipelines/pipeline-list/pipeline-list.component.html [hunk 14]
```diff
@@ -1,6 +1,17 @@
 <app-container>
   <h1 class='theme-primary-fg' i18n>Processing Tasks</h1>
   <div class='controls'>
+    <div>
+      <mat-form-field appearance="fill">
+        <mat-label>Sort</mat-label>
+        <select [(ngModel)]='sortBySelected'>
+          <option value = 'last_created'></option>
```
Comment (akariv) [inline ui/projects/etl-server/src/lib/pipelines/pipeline-list/pipeline-list.component.html:8]: why do we need this empty option? Why not set the default to created_at?

File: ui/projects/etl-server/src/lib/pipelines/pipeline-list/pipeline-list.component.ts [hunk 15]
```diff
@@ -11,6 +11,8 @@ import { RolesService } from '../../roles.service';
 })
 export class PipelineListComponent implements OnInit {
 
+  sortByField = '';
```
Comment (akariv) [inline ui/projects/etl-server/src/lib/pipelines/pipeline-list/pipeline-list.component.ts:14]: Here's the place where the default value should be set.

File: ui/projects/etl-server/src/lib/pipelines/pipeline-list/pipeline-list.component.ts [hunk 16]
```diff
@@ -47,11 +49,24 @@ export class PipelineListComponent implements OnInit {
...
     });
     this.api.queryPipelines();
   }
+  get sortBySelected() {
+    return this.sortByField;
+  }
+
+  set sortBySelected(value) {
+    this.sortByField = value;
+    if (this.sortByField !== 'last_created') {
```
Comment (akariv) [inline ui/projects/etl-server/src/lib/pipelines/pipeline-list/pipeline-list.component.ts:64]: Still not sure why this if is needed and why the `last_created` value was added here.

File: ui/projects/etl-server/src/lib/pipelines/pipeline-list/pipeline-list.component.ts [hunk 17]
```diff
@@ -47,11 +49,24 @@ export class PipelineListComponent implements OnInit {
...
+      this.pipelineSections.sort((a,b) => a.item[this.sortByField] > b.item[this.sortByField] ? 1 : -1);
```
Comment (akariv) [inline ui/projects/etl-server/src/lib/pipelines/pipeline-list/pipeline-list.component.ts:65]: Just a reminder - you need to sort the actual pipelines and then recalculate the sections.
You might need to save the list of pipelines that we receive from the API so you can re-sort them here.

**Review by akariv — CHANGES_REQUESTED** [review]

File: ui/projects/etl-server/src/lib/pipelines/pipeline-list/pipeline-list.component.ts [hunk 18]
```diff
@@ -1,4 +1,6 @@
 import { Component, OnInit } from '@angular/core';
+import { relativeTimeThreshold } from 'moment';
```
Comment (akariv) [inline ui/projects/etl-server/src/lib/pipelines/pipeline-list/pipeline-list.component.ts:2]: Looks like these imports got here by mistake?

**Review by akariv — CHANGES_REQUESTED** [review]

File: ui/projects/etl-server/src/lib/pipelines/pipeline-list/pipeline-list.component.ts [hunk 19]
```diff
@@ -49,11 +51,22 @@ export class PipelineListComponent implements OnInit {
...
     });
     this.api.queryPipelines();
   }
+  get sortBySelected() {
+    return this.sortByField;
+  }
+
+  set sortBySelected(value) {
+    this.sortByField = value;
+    this.pipelines.sort((a,b) => a[this.sortByField] > b[this.sortByField] ? 1 : -1);
```
Comment (akariv) [inline ui/projects/etl-server/src/lib/pipelines/pipeline-list/pipeline-list.component.ts:67]: Just thought of something - when sorting by update date, we want the one most recently updated to come first (like in gmail).
When sorting by name, we want an A-Z sorting.
From my reading of the code it's currently not like that, correct?
